docs/compactor.md

```diff
@@ -0,0 +1,93 @@
+
+**How the solution works:** The newly introduced compact command does not disturb the running ETCD while capturing the compacted snapshot. The command is designed to run potentially separately (from the main ETCD process/container/pod). ETCD Druid may take care of starting the newly introduced compact command as a separate job (scheduled periodically and also on-demand if required) based on some user defined parameter.
+
+### **Points to take care while saving the compacted snapshot:**
+As compacted snapshot and the existing periodic full snapshots are taken by different processes running in different pods but accessing same store to save the snapshots, some problems may arise:
+1.	When uploading the compacted snapshot to the snapstore, there is the problem of how does the restorer know when to start using the newly compacted snapshot. This communication needs to be atomic.
+2.	With a regular schedule for compaction that happens potentially separately from the main etcd pod, is there a need for regular scheduled full snapshots anymore?
+
+#### **How to swap full snapshot with compacted snapshot atomically**
+
+Currently, full snapshots and the subsequent delta snapshots are grouped under same prefix path in the snapstore. When a full snapshot is created, it is placed under a prefix/directory with the name comprising of timestamp. Then subsequent delta snapshots are also pushed into the same directory. Thus each prefix/directory contains a single full snapshot and the subsequent delta snapshots. So far, it is the job of ETCDBR to start main ETCD process and snapshotter process which takes full snapshot and delta snapshot periodically. But as per our proposal, compaction will be running as parallel process to main ETCD process and snapshotter process. So we can't reliably co-ordinate between the processes to achieve switching to the compacted snapshot as the base snapshot atomically.```
+
+**Current Directory Structure**
+```yaml
+- Backup-192345
+    - Full-Snapshot-192345
+    - Incremental-Snapshot-192355
+    - Incremental-Snapshot-192365
+    - Incremental-Snapshot-192375
+- Backup-192789
+    - Full-Snapshot-192789
+    - Incremental-Snapshot-192799
+    - Incremental-Snapshot-192809
+    - Incremental-Snapshot-192819
+```
+
+To solve the problem, proposal is:
+1. ETCDBR will take the first full snapshot after it starts main ETCD Process and snapshotter process. But after taking the first full snapshot, snapshotter will only continue taking delta snapshots but no further full snapshots.
+2. Flatten the directory structure of backup folder. Save all the full snapshots, delta snapshots and compacted snapshots under same directory/prefix. Also, name snapshots after revision number instead of timestamps. Restorer will restore from compacted snapshots that have highest revision number in the name and delta snapshots that have higher revision numbers in name than the comapcted full snapshot.
```

Comment:
> Also, name snapshots after revision number instead of timestamps

I guess this is now out-of-date. We didn't change the snapshot file names but only the sorting criteria, right?

Reply:
Yes, but sorting criteria will be based on revision number present in the name.